### PR TITLE
feat(owlbot): CLI for creating copy jobs in reaction to updates to googleapis-gen

### PR DESCRIPTION
**Pull request by bcoe**
**see inline comments for description of moving parts.**

### CODE REVIEW

File: packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      const octokit = await getAuthenticatedOctokit(token.token);
+      admin.initializeApp({
+        credential: admin.credential.applicationDefault(),
+        projectId: argv['firestore-project'],
+      });
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const db = admin.firestore();
+      const configStore = new FirestoreConfigsStore(db!);
+
+      const shas = await getLastNCommits(argv.repo, octokit, 1);
```
Comment (bcoe) [inline packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:123]: I was figuring we'd start by looking at the last commit until we've bootstrapped the repository, i.e., have a history of commits, at which point we can pull some reasonable number of previous commits to look back in time, and ensure we're up-to-date.

File: packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      console.info(shas);
+      // TODO: stop hardcoding this once we've tested the system:
+      const sha = 'b32470c59b90b1bb22c29d285474c19870d6f3f1';
```
Comment (bcoe) [inline packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:126]: picked a sha with a huge number of files for testing.

File: packages/owl-bot/src/owl-bot.ts [hunk 3]
```diff
@@ -157,4 +157,30 @@ export = (privateKey: string | undefined, app: Probot, db?: Db) => {
...
+  /*
+  const GOOGLE_API_GEN_REPO = 'googleapis/googleapis-gen';
+  app.on('push', async context => {
+    const configStore = new FirestoreConfigsStore(db!);
+    logger.info(`push event for ${context.payload.repository.full_name}`);
+    if (context.payload.repository.full_name === GOOGLE_API_GEN_REPO) {
+      const pathSet: Set<string> = new Set();
+      for (const commit of context.payload.commits) {
+        const allPaths = [
+          ...commit.added,
```
Comment (bcoe) [inline packages/owl-bot/src/owl-bot.ts:170]: TODO: instead enqueue background task.

**Review by bcoe — COMMENTED** [review]
added some notes about implementation.

File: packages/owl-bot/cloud-build/enqueue-copy-jobs.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+steps:
+  # Clone googleapis-gen, so that we can list modified files:
+  - name: 'gcr.io/cloud-builders/git'
+    args:
```
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/cloud-build/enqueue-copy-jobs.yaml:4]: Please add the note about Github's 3000 file limit to the comment on line 2.

File: packages/owl-bot/cloud-build/enqueue-copy-jobs.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      - https://github.com/googleapis/googleapis-gen.git
+      # - --depth=50
+  # Output the GitHub private key to disk for benefit of CLI:
+  - name: gcr.io/cloud-builders/gcloud
+    entrypoint: /bin/bash
+    args: ['-c', 'echo "$$PRIVATE_KEY" > private-key.pem']
+    dir: 'googleapis-gen'
+    env:
+      - 'PRIVATE_KEY=$_PRIVATE_KEY'
+  # Run the "enqueue-copy-tasks" CLI task:
```
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/cloud-build/enqueue-copy-jobs.yaml:16]: Replace the comment on line 16 with this comment.

File: packages/owl-bot/cloudbuild.yaml [hunk 6]
```diff
@@ -36,3 +36,10 @@ steps:
       - "$_KEY_RING"
       - "$_FUNCTION_REGION"
       - "nodejs12"
+  
+  # Build the owlbot command line, which is used in the schedule PRs build:
+  - name: 'gcr.io/cloud-builders/docker'
+    id: "build-docker"
```
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/cloudbuild.yaml:42]: Insert this comment before line 41.

File: packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+      })
+      .option('repo', {
+        describe: 'repository to monitor for changes',
+        type: 'string',
+        default: 'googleapis/googleapis-gen',
+      })
+      .option('git-path', {
+        describe: 'path on disk of repo to monitor for changes',
+        type: 'string',
+        demand: true,
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:72]: Not sure what I should pass here.  Could you specify a default, show an example, or describe in more detail?

File: packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        type: 'string',
+        default: 'repo-automation-bots',
+      })
+      .option('firestore-project', {
+        describe: 'project used for firestore database',
+        type: 'string',
+        default: 'repo-automation-bots-metrics',
+      })
+      .option('queue', {
+        describe: 'queue to publish PR update jobs to',
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:85]: _pubsub_ queue to publish PR update jobs to

File: packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        type: 'string',
+        default: 'projects/repo-automation-bots/topics/owlbot-prs',
+      })
+      .option('trigger', {
+        describe:
+          'enqueue copy job trigger (if this should be run in background)',
+        type: 'string',
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:92]: Not sure what I should pass here. Could you specify a default, show an example, or describe in more detail?

File: packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+      const files = await getFilesModifiedBySha(argv['git-path'], sha);
+      logger.info(`found ${files.length} files changed`);
+      const repos = await configStore.findReposAffectedByFileChanges(files);
+      for (const repo of repos) {
+        let messageId = await configStore.findPubsubMessageIdForCopyTask(
+          repo,
+          sha
+        );
+        if (messageId) {
+          logger.info(`${repo} has copy job ${messageId}`);
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:136]: Let's include the decision we're making in this log line.  For example,
`${repo} already has copy job ${messageId}, so I won't create another one.`

File: packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+          continue;
+        }
+        logger.info(`enqueue ${argv.repo}`);
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:139]: Can we embellish this log line too?

File: packages/owl-bot/src/core.ts [hunk 12]
```diff
@@ -359,14 +402,104 @@ export async function getFileContent(
...
+    cwd: path,
+    // Handle 100,000+ files changing:
+    maxBuffer: 1024 * 1024 * 512,
+  });
+  if (out.stderr) throw Error(out.stderr);
+  const filesRaw = out.stdout.trim();
+  const files = [];
+  // We walk the output in reverse, since the file list is shown at the end
+  // of git show:
+  for (const file of filesRaw.split(/\r?\n/).reverse()) {
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/core.ts:422]: Nice.  Simple strategy yields simple code.

File: packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        describe: 'GitHub AppID',
+        type: 'number',
+        demand: true,
+      })
+      .option('installation', {
+        describe: 'installation ID for GitHub app',
+        type: 'number',
+        demand: true,
+      })
+      .option('repo', {
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:64]: To keep the naming consistent with other commands, could we call this the source repo?

File: packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts (hunk 1 above)
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:123]: What's the goal?

File: packages/owl-bot/src/database.ts [hunk 14]
```diff
@@ -161,7 +161,9 @@ export class FirestoreConfigsStore implements ConfigsStore, CopyTasksStore {
     // to optimize if performance becomes a problem.
     const snapshot = await this.db.collection(this.yamls).get();
     const result: string[] = [];
+    let i = 0;
     snapshot.forEach(doc => {
+      i++;
       const configs = doc.data() as Configs | undefined;
       match_loop: for (const copy of configs?.yaml?.['copy-dirs'] ?? []) {
         const mm = newMinimatchFromSource(copy.source);
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/database.ts:169]: Oh dear.  This code should be using regular expressions instead of minimatch.  I'll prepare a PR.

**Review by SurferJeffAtGoogle — CHANGES_REQUESTED** [review]

File: packages/owl-bot/src/database.ts (hunk 14 above)
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/src/database.ts:169]: Nevermind.  The code that switches this to regular expressions is in my pending PR.

File: packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts (hunk 1 above)
Comment (bcoe) [reply packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:123]: I was thinking about the task we schedule that walks commits every hour or so, looking for any we missed. It seemed like if we pick a reasonable number of commits to look back in time, e.g., 25:

* we can backfill 25 commits to make sure that they're in the DB.
* always check 25 commits when we're ensuring we're caught up.

Comment (bcoe) [reply packages/owl-bot/src/bin/commands/enqueue-copy-tasks.ts:123]: I've refactored this logic to use an infinite async iterator instead.

File: packages/owl-bot/src/core.ts [hunk 15]
```diff
@@ -359,14 +362,67 @@ export async function getFileContent(
...
+    // There will be a blank line between the commit message and the
+    // files list, we use this as a stopping point:
+    if (file === '') break;
+    files.push(file);
+  }
+  return files;
+}
+
+/**
+ * Returns the last 'n' commits made to a repository.
```
Comment (bcoe) [inline packages/owl-bot/src/core.ts:392]:
```suggestion
 * Returns an iterator that returns the most recent commits added to a repository.
```

File: packages/owl-bot/test/fixtures/push.json [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+      "id": 16785467,
+      "node_id": "MDEyOk9yZ2FuaXphdGlvbjE2Nzg1NDY3",
+      "avatar_url": "https://avatars.githubusercontent.com/u/16785467?v=4",
+      "gravatar_id": "",
+      "url": "https://api.github.com/users/googleapis",
+      "html_url": "https://github.com/googleapis",
+      "followers_url": "https://api.github.com/users/googleapis/followers",
+      "following_url": "https://api.github.com/users/googleapis/following{/other_user}",
+      "gists_url": "https://api.github.com/users/googleapis/gists{/gist_id}",
+      "starred_url": "https://api.github.com/users/googleapis/starred{/owner}{/repo}",
```
Comment (bcoe) [inline packages/owl-bot/test/fixtures/push.json:25]: This fixture immediately becomes useful once we implement the real-time endpoints, I figured let's just leave it.

**Review by SurferJeffAtGoogle — APPROVED** [review]

**Pull request merged**